Added menu and menu link traits.

richardgaunt opened the pull request:
## Feature Request

I wanted to create menus via Behat in D8, menu links which can be nested and enabled/ disabled to test custom site functionality.

## What has changed
1. Updated menu trait to include creating of menus
2. Created menu_link_content trait to allow adding of menu items to menus
3. Created tests verifying the above features

AlexSkrypnyk commented on src/D8/MenuLinkContentTrait.php:
```diff
@@ -0,0 +1,63 @@
+   *
+   * Provide menu data in the following format:
+   *
+   * | id    | uuid   | title        | description     | menu_name | enabled |
```
please fix the format of this comment. it should be set in a single line to show the table correctly

AlexSkrypnyk replied:
The only required item should be the link `title` and the rest should be optional ones

AlexSkrypnyk commented on src/D8/MenuLinkContentTrait.php:
```diff
@@ -0,0 +1,63 @@
+   * @Given menu_links:
+   */
+  public function menuLinksCreate(TableNode $table) {
+    $this->menuLinksDelete($table);
```
please remove this implicit behaviour. removal of links should happen explicitly.

richardgaunt replied:
Are ok with writing an AfterScenario method to handle cleanup?

AlexSkrypnyk commented on src/D8/MenuLinkContentTrait.php:
```diff
@@ -0,0 +1,63 @@
+   *
+   * Only the 'id' field is required.
+   *
+   * @Given menu_links:
```
Please update the step definition to include the menu name
`@Given "mymenu" menu links:`

AlexSkrypnyk commented on src/D8/MenuLinkContentTrait.php:
```diff
@@ -0,0 +1,63 @@
+   *
+   * @Given menu_links:
+   */
+  public function menuLinksCreate(TableNode $table) {
```
```
 public function menuLinksCreate($menu_name, TableNode $table) {
```

AlexSkrypnyk commented on src/D8/MenuLinkContentTrait.php:
```diff
@@ -0,0 +1,63 @@
+   * | ...   | ...    | ...          | ...             | ...       | ...     |
+   * ...                     | ...                      |
+   *
+   * Only the 'id' field is required.
```
this cannot be the case by default - the ID is an internal field and not only should not be required, but should be discouraged to be used as menu ID is an auto-increment fields and Drupal should be handling the ID creation internally.

AlexSkrypnyk commented on src/D8/MenuLinkContentTrait.php:
```diff
@@ -0,0 +1,63 @@
+   *
+   * Provide menu data in the following format:
+   *
+   * | id        |
```
using IDs is a very narrow use case. usually, links are referred by their Titles. Please implement a lookup system by title. If the `id` or `link` is provided - then lookup should be using menu items that match all of the criteria

richardgaunt replied:
Thanks Alex - this is solid feedback, I've adapted the steps to work on menu titles, and menu link titles so that IDs are not required and not needed by default.

AlexSkrypnyk commented on src/D8/MenuLinkContentTrait.php:
```diff
@@ -0,0 +1,63 @@
+   * | 9991      |
+   * | ...       |
+   *
+   * @Given no menu_links:
```
```
Given no "mymenu" menu links:
```

richardgaunt replied:
Thanks this is a nice enhancement - I've adopted

AlexSkrypnyk commented on src/D8/MenuLinkContentTrait.php:
```diff
@@ -0,0 +1,63 @@
+   *
+   * @Given no menu_links:
+   */
+  public function menuLinksDelete(TableNode $table) {
```
`public function menuLinksDelete($menu_name, TableNode $table) {`

AlexSkrypnyk commented on src/D8/MenuLinkContentTrait.php:
```diff
@@ -0,0 +1,63 @@
+<?php
```
please move all this into `MenuTrait` as these traits are grouped by functionality rather than Drupal structures.

AlexSkrypnyk commented on src/SelectTrait.php:
```diff
@@ -39,4 +39,38 @@ public function selectShouldNotHaveOption($select, $option) {
+   * @Then select :select should have option :option selected
+   */
+  public function selectShouldHaveOptionSelected($select, $option) {
+    $selectElement = $this->getSession()->getPage()->findField($select);
```
please refactor to use `selectShouldHaveOption()` (add returning of the element to `selectShouldHaveOption()` ). and then use `getAttribute()`.

AlexSkrypnyk commented on tests/behat/features/d8.menu.feature:
```diff
@@ -0,0 +1,44 @@
+@d8 @wip
```
please remove `@wip`

AlexSkrypnyk reviewed (changes requested):
@richardgaunt 

Thank you so much for contributing!

There are some changes that need to be added to keep this new functionality in-line with the rest of traits. Please see my comments.

richardgaunt commented:
Thanks @AlexSkrypnyk I've updated based on your feedback, thank you for taking the time to give the review.
I have moved all the code within MenuTrait, refactored to use title and labels to reference other created entities and made other minor adjustments on your feedback.

AlexSkrypnyk commented:
@richardgaunt Thank you so much for working on this!
